pkg/restore/cdclog/decoder.go

```diff
@@ -0,0 +1,273 @@
+}
+
+func (c column) toDatum() (types.Datum, error) {
+	var (
+		val interface{}
+		err error
+	)
+
+	switch c.Type {
+	case types.KindInt64, types.KindUint64:
+		val, err = c.Value.(json.Number).Int64()
+		if err != nil {
+			return types.Datum{}, errors.Trace(err)
+		}
+	case types.KindFloat32, types.KindFloat64:
+		val, err = c.Value.(json.Number).Float64()
+		if err != nil {
+			return types.Datum{}, errors.Trace(err)
+		}
+	default:
+		val = c.Value
+	}
+	return types.NewDatum(val), nil
+}
+
+func formatColumnVal(c column) column {
+	switch c.Type {
+	case mysql.TypeTinyBlob, mysql.TypeMediumBlob,
+		mysql.TypeLongBlob, mysql.TypeBlob:
+		if s, ok := c.Value.(string); ok {
+			var err error
+			c.Value, err = base64.StdEncoding.DecodeString(s)
+			if err != nil {
+				log.Fatal("invalid column value, please report a bug", zap.Any("col", c), zap.Error(err))
+			}
+		}
+	case mysql.TypeBit:
+		if s, ok := c.Value.(json.Number); ok {
+			intNum, err := s.Int64()
+			if err != nil {
+				log.Fatal("invalid column value, please report a bug", zap.Any("col", c), zap.Error(err))
+			}
+			c.Value = uint64(intNum)
+		}
+	}
+	return c
+}
+
+type messageKey struct {
+	Ts        uint64 `json:"ts"`
+	Schema    string `json:"scm,omitempty"`
+	Table     string `json:"tbl,omitempty"`
+	RowID     int64  `json:"rid,omitempty"`
+	Partition *int64 `json:"ptn,omitempty"`
+}
+
+// Encode the messageKey.
+func (m *messageKey) Encode() ([]byte, error) {
+	return json.Marshal(m)
+}
+
+// Decode the messageKey.
+func (m *messageKey) Decode(data []byte) error {
+	return json.Unmarshal(data, m)
+}
+
+// MessageDDL represents the ddl changes.
+type MessageDDL struct {
+	Query string             `json:"q"`
+	Type  timodel.ActionType `json:"t"`
+}
+
+// Encode the DDL message.
+func (m *MessageDDL) Encode() ([]byte, error) {
+	return json.Marshal(m)
+}
+
+// Decode the DDL message.
+func (m *MessageDDL) Decode(data []byte) error {
+	return json.Unmarshal(data, m)
+}
+
+// MessageRow represents the row changes in same commit ts.
+type MessageRow struct {
+	Update     map[string]column `json:"u,omitempty"`
+	PreColumns map[string]column `json:"p,omitempty"`
+	Delete     map[string]column `json:"d,omitempty"`
+}
+
+// Encode the Row message.
+func (m *MessageRow) Encode() ([]byte, error) {
+	return json.Marshal(m)
+}
+
+// Decode the Row message.
+func (m *MessageRow) Decode(data []byte) error {
+	decoder := json.NewDecoder(bytes.NewReader(data))
+	decoder.UseNumber()
+	err := decoder.Decode(m)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	for colName, column := range m.Update {
+		m.Update[colName] = formatColumnVal(column)
+	}
+	for colName, column := range m.Delete {
+		m.Delete[colName] = formatColumnVal(column)
+	}
+	for colName, column := range m.PreColumns {
+		m.PreColumns[colName] = formatColumnVal(column)
+	}
+	return nil
+}
+
+// SortItem represents a DDL item or Row changed item.
+type SortItem struct {
+	ItemType ItemType
+	Meta     interface{}
```

Comment:
Seems this field stores the "content" of this `SortItem`. Hence the name `Meta` is someway misleading, how about renaming it to `Message` or `Data`?

Reply:
Done, move it to `Data`